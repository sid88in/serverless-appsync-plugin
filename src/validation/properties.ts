import { timeUnits } from '../utils.js';

export const name = { type: 'string' };
// Depends on auth
export const authentication = { $ref: '#/definitions/auth' };
export const schema = {
  anyOf: [
    {
      type: 'string',
    },
    {
      type: 'array',
      items: { type: 'string' },
    },
  ],
  errorMessage: 'must be a valid schema config',
};
//Depends on stringOrIntrinsicFunction
export const domain = {
  type: 'object',
  properties: {
    enabled: { type: 'boolean' },
    useCloudFormation: { type: 'boolean' },
    retain: { type: 'boolean' },
    name: {
      type: 'string',
      pattern: '^([a-z][a-z0-9+-]*\\.)+[a-z][a-z0-9]*$',
      errorMessage: 'must be a valid domain name',
    },
    certificateArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    hostedZoneId: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    hostedZoneName: {
      type: 'string',
      pattern: '^([a-z][a-z0-9+-]*\\.)+[a-z][a-z0-9]*\\.$',
      errorMessage:
        'must be a valid zone name. Note: you must include a trailing dot (eg: `example.com.`)',
    },
    route53: { type: 'boolean' },
  },
  required: ['name'],
  if: {
    anyOf: [
      {
        not: { properties: { useCloudFormation: { const: false } } },
      },
      { not: { required: ['useCloudFormation'] } },
    ],
  },
  then: {
    anyOf: [{ required: ['certificateArn'] }, { required: ['hostedZoneId'] }],
    errorMessage:
      'when using CloudFormation, you must provide either certificateArn or hostedZoneId.',
  },
};
export const xrayEnabled = { type: 'boolean' };
export const visibility = {
  type: 'string',
  enum: ['GLOBAL', 'PRIVATE'],
  errorMessage: 'must be "GLOBAL" or "PRIVATE"',
};
export const introspection = { type: 'boolean' };
export const queryDepthLimit = { type: 'integer', minimum: 1, maximum: 75 };
export const resolverCountLimit = {
  type: 'integer',
  minimum: 1,
  maximum: 1000,
};
// Depends on substitutions
export const substitutions = { $ref: '#/definitions/substitutions' };
// Depends on environment
export const environment = { $ref: '#/definitions/environment' };
// Depends on stringOrIntrinsicFunction and wafRule
export const waf = {
  type: 'object',
  properties: {
    enabled: { type: 'boolean' },
  },
  if: {
    required: ['arn'],
  },
  then: {
    properties: {
      arn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    },
  },
  else: {
    properties: {
      name: { type: 'string' },
      defaultAction: {
        type: 'string',
        enum: ['Allow', 'Block'],
        errorMessage: "must be 'Allow' or 'Block'",
      },
      description: { type: 'string' },
      rules: {
        type: 'array',
        items: { $ref: '#/definitions/wafRule' },
      },
    },
    required: ['rules'],
  },
};
export const tags = {
  type: 'object',
  additionalProperties: { type: 'string' },
};
export const caching = {
  type: 'object',
  properties: {
    enabled: { type: 'boolean' },
    behavior: {
      type: 'string',
      enum: ['FULL_REQUEST_CACHING', 'PER_RESOLVER_CACHING'],
      errorMessage:
        "must be one of 'FULL_REQUEST_CACHING', 'PER_RESOLVER_CACHING'",
    },
    type: {
      enum: [
        'SMALL',
        'MEDIUM',
        'LARGE',
        'XLARGE',
        'LARGE_2X',
        'LARGE_4X',
        'LARGE_8X',
        'LARGE_12X',
      ],
      errorMessage:
        "must be one of 'SMALL', 'MEDIUM', 'LARGE', 'XLARGE', 'LARGE_2X', 'LARGE_4X', 'LARGE_8X', 'LARGE_12X'",
    },
    ttl: { type: 'integer', minimum: 1, maximum: 3600 },
    atRestEncryption: { type: 'boolean' },
    transitEncryption: { type: 'boolean' },
  },
  required: ['behavior'],
};
// Depends on auth
export const additionalAuthentications = {
  type: 'array',
  items: { $ref: '#/definitions/auth' },
};
// Depends on wafRule
export const apiKeys = {
  type: 'array',
  items: {
    if: { type: 'object' },
    then: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        description: { type: 'string' },
        expiresAfter: {
          type: ['string', 'number'],
          pattern: `^(\\d+)(${Object.keys(timeUnits).join('|')})?$`,
          errorMessage: 'must be a valid duration.',
        },
        expiresAt: {
          type: 'string',
          format: 'date-time',
          errorMessage: 'must be a valid date-time',
        },
        wafRules: {
          type: 'array',
          items: { $ref: '#/definitions/wafRule' },
        },
      },
      required: ['name'],
    },
    else: {
      type: 'string',
    },
  },
};
// Depends on stringOrIntrinsicFunction
export const logging = {
  type: 'object',
  properties: {
    roleArn: { $ref: '#/definitions/stringOrIntrinsicFunction' },
    level: {
      type: 'string',
      enum: ['ALL', 'INFO', 'DEBUG', 'ERROR', 'NONE'],
      errorMessage: "must be one of 'ALL', 'INFO', 'DEBUG', 'ERROR' or 'NONE'",
    },
    retentionInDays: { type: 'integer' },
    excludeVerboseContent: { type: 'boolean' },
    enabled: { type: 'boolean' },
  },
  required: ['level'],
};
// Depends on dataSourceConfig
export const dataSources = {
  oneOf: [
    {
      type: 'object',
      additionalProperties: { $ref: '#/definitions/dataSourceConfig' },
    },
    {
      type: 'array',
      items: {
        type: 'object',
        additionalProperties: { $ref: '#/definitions/dataSourceConfig' },
      },
    },
  ],
  errorMessage: 'contains invalid data source definitions',
};
// Depends on resolverConfigMap
export const resolvers = {
  oneOf: [
    { $ref: '#/definitions/resolverConfigMap' },
    {
      type: 'array',
      items: { $ref: '#/definitions/resolverConfigMap' },
    },
  ],
  errorMessage: 'contains invalid resolver definitions',
};
// Depends on pipelineFunctionConfigMap
export const pipelineFunctions = {
  oneOf: [
    {
      $ref: '#/definitions/pipelineFunctionConfigMap',
    },
    {
      type: 'array',
      items: {
        $ref: '#/definitions/pipelineFunctionConfigMap',
      },
    },
  ],
  errorMessage: 'contains invalid pipeline function definitions',
};
// Depends on stringOrIntrinsicFunction
export const esbuild = {
  oneOf: [
    {
      type: 'object',
    },
    { const: false },
  ],
  errorMessage: 'must be an esbuild config object or false',
};
export const apiId = { $ref: '#/definitions/stringOrIntrinsicFunction' };
