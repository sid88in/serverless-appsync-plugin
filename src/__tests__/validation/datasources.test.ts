import { validateConfig } from '../../validation';
import { basicConfig } from '../basicConfig';

describe('Basic', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'AMAZON_DYNAMODB',
              description: 'My Dynamo Datasource',
              config: {
                tableName: 'myTable',
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              myDynamoSource1: {
                type: 'AMAZON_DYNAMODB',
                description: 'My Dynamo Datasource',
                config: {
                  tableName: 'myTable',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });

  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Invalid Datasource',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'Foo',
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('DynamoDB', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'AMAZON_DYNAMODB',
              config: {
                tableName: 'myTable',
              },
            },
            myDynamoSource2: {
              type: 'AMAZON_DYNAMODB',
              config: {
                tableName: { Ref: 'MyTable' },
                region: { 'Fn::Sub': '${AWS::Region}' },
                serviceRoleArn: { 'Fn::GetAtt': 'MyRole.Arn' },
              },
            },
            myDynamoSource3: {
              type: 'AMAZON_DYNAMODB',
              config: {
                tableName: 'myTable',
                useCallerCredentials: true,
                region: 'us-east-2',
                serviceRoleArn: 'arn:',
                iamRoleStatements: [
                  {
                    Effect: 'Allow',
                    Action: ['DynamoDB:PutItem'],
                    Resource: ['arn:dynamodb:'],
                  },
                ],
                versioned: true,
                deltaSyncConfig: {
                  deltaSyncTableName: 'deltaSyncTable',
                  baseTableTTL: 60,
                  deltaSyncTableTTL: 60,
                },
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              myDynamoSource1: {
                type: 'AMAZON_DYNAMODB',
                config: {
                  tableName: 'myTable',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });

  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Missing config',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'AMAZON_DYNAMODB',
            },
          },
        },
      },
      {
        name: 'Empty config',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'AMAZON_DYNAMODB',
              config: {},
            },
          },
        },
      },
      {
        name: 'Invalid config',
        config: {
          dataSources: {
            myDynamoSource1: {
              type: 'AMAZON_DYNAMODB',
              config: {
                tableName: 123,
                useCallerCredentials: 'foo',
                region: 123,
                serviceRoleArn: 456,
                iamRoleStatements: [{}],
                versioned: 'bar',
                deltaSyncConfig: {
                  baseTableTTL: '123',
                  deltaSyncTableTTL: '456',
                },
              },
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('Lambda', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {
                functionName: 'myTable',
              },
            },
            myLambda2: {
              type: 'AWS_LAMBDA',
              config: {
                functionArn: 'arn:lambda',
                serviceRoleArn: 'arn:iam',
              },
            },
            myLambda3: {
              type: 'AWS_LAMBDA',
              config: {
                functionArn: { Ref: 'MyLambda' },
                serviceRoleArn: { Ref: 'MyRole' },
              },
            },
            myLambda4: {
              type: 'AWS_LAMBDA',
              config: {
                functionName: 'myLambda',
                iamRoleStatements: [
                  {
                    Effect: 'Allow',
                    Action: ['lambda:invokeFunction'],
                    Resource: ['arn:lambda:'],
                  },
                ],
              },
            },
            myLambda5: {
              type: 'AWS_LAMBDA',
              config: {
                function: {
                  handler: 'index.handler',
                  timeout: 30,
                },
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              myLambda1: {
                type: 'AWS_LAMBDA',
                config: {
                  functionName: 'myTable',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });
  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Missing config',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
            },
          },
        },
      },
      {
        name: 'Empty config',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {},
            },
          },
        },
      },
      {
        name: 'Invalid config',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {
                tableName: 123,
              },
            },
          },
        },
      },
      {
        name: 'Invalid functionName',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {
                functionName: 123,
              },
            },
          },
        },
      },
      {
        name: 'Invalid functionArn',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {
                functionArn: 123,
              },
            },
          },
        },
      },
      {
        name: 'Invalid embedded function',
        config: {
          dataSources: {
            myLambda1: {
              type: 'AWS_LAMBDA',
              config: {
                function: 'myFunction',
              },
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('RelationalDb', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            rds1: {
              type: 'RELATIONAL_DATABASE',
              config: {
                dbClusterIdentifier: 'myClisterId',
                awsSecretStoreArn: 'aws:arn:',
              },
            },
            rds2: {
              type: 'RELATIONAL_DATABASE',
              config: {
                dbClusterIdentifier: 'myClisterId',
                relationalDatabaseSourceType: 'RDS_HTTP_ENDPOINT',
                region: 'us-east-1',
                dataBaseName: 'myDatabase',
                schema: '',
                awsSecretStoreArn: { Reg: 'MySecretStore' },
                iamRoleStatements: [
                  {
                    Effect: 'Allow',
                    Action: ['rds-data:GetItems'],
                    Resource: ['aws:arn:rds:'],
                  },
                ],
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              rds1: {
                type: 'RELATIONAL_DATABASE',
                config: {
                  dbClusterIdentifier: 'myClisterId',
                  awsSecretStoreArn: 'aws:arn:',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });

  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Missing config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
            },
          },
        },
      },
      {
        name: 'Empty config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
              config: {},
            },
          },
        },
      },
      {
        name: 'Invalid config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
              config: {
                endpoint: 123,
                authorizationConfig: {
                  authorizationType: 'FOO',
                  awsIamConfig: {
                    signingRegion: 123,
                    signingServiceName: 456,
                  },
                },
              },
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('HTTP', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
              config: {
                endpoint: 'https://api.example.com',
              },
            },
            http2: {
              type: 'HTTP',
              config: {
                endpoint: { 'Fn::GetAtt': ['MyEndpoint', 'Arn'] },
                authorizationConfig: {
                  authorizationType: 'AWS_IAM',
                  awsIamConfig: {
                    signingRegion: 'us-east-1',
                    signingServiceName: 'AppSync',
                  },
                },
              },
            },
            http3: {
              type: 'HTTP',
              config: {
                endpoint: 'https://api.example.com',
                serviceRoleArn: { Ref: 'MyRole' },
              },
            },
            http5: {
              type: 'HTTP',
              config: {
                endpoint: 'https://api.example.com',
                iamRoleStatements: [
                  {
                    Effect: 'Allow',
                    Action: ['lambda:invokeFunction'],
                    Resource: ['arn:lambda:'],
                  },
                ],
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              http1: {
                type: 'HTTP',
                config: {
                  endpoint: 'https://api.example.com',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });
  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Missing config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
            },
          },
        },
      },
      {
        name: 'Empty config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
              config: {},
            },
          },
        },
      },
      {
        name: 'Invalid config',
        config: {
          dataSources: {
            http1: {
              type: 'HTTP',
              config: {
                endpoint: 123,
                authorizationConfig: {
                  authorizationType: 'FOO',
                  awsIamConfig: {
                    signingRegion: 123,
                    signingServiceName: 456,
                  },
                },
              },
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('OpenSearch', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            openSearch1: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                endpoint: 'https://api.example.com',
              },
            },
            openSearch2: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                endpoint: { 'Fn::GetAtt': ['MyEndpoint', 'Arn'] },
              },
            },
            openSearch3: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                domain: '123',
              },
            },
            openSearch4: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                endpoint: 'https://api.example.com',
                region: 'us-east-1',
                serviceRoleArn: 'aws:arn:iam',
              },
            },
            openSearch5: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                endpoint: 'https://api.example.com',
                iamRoleStatements: [
                  {
                    Effect: 'Allow',
                    Action: ['lambda:invokeFunction'],
                    Resource: ['arn:lambda:'],
                  },
                ],
              },
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              openSearch1: {
                type: 'AMAZON_OPENSEARCH_SERVICE',
                config: {
                  endpoint: 'https://api.example.com',
                },
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });
  describe('Invalid', () => {
    const assertions = [
      {
        name: 'Missing config',
        config: {
          dataSources: {
            openSearch1: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
            },
          },
        },
      },
      {
        name: 'Empty config',
        config: {
          dataSources: {
            openSearch1: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {},
            },
          },
        },
      },
      {
        name: 'Invalid config',
        config: {
          dataSources: {
            openSearch1: {
              type: 'AMAZON_OPENSEARCH_SERVICE',
              config: {
                endpoint: 123,
                region: 456,
              },
            },
          },
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate: ${config.name}`, () => {
        expect(function () {
          validateConfig({
            ...basicConfig,
            ...config.config,
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});

describe('None', () => {
  describe('Valid', () => {
    const assertions = [
      {
        name: 'Valid config',
        config: {
          dataSources: {
            none: {
              type: 'NONE',
            },
          },
        },
      },
      {
        name: 'Valid config, as array of maps',
        config: {
          dataSources: [
            {
              openSearch1: {
                type: 'NONE',
              },
            },
          ],
        },
      },
    ];

    assertions.forEach((config) => {
      it(`should validate a ${config.name}`, () => {
        expect(validateConfig({ ...basicConfig, ...config.config })).toBe(true);
      });
    });
  });
});
