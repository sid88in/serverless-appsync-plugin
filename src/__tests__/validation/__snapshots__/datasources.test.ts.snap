// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Invalid should validate: Invalid Datasource 1`] = `
"/dataSources/myDynamoSource1/type: must be one of AMAZON_DYNAMODB, AMAZON_OPENSEARCH_SERVICE, AWS_LAMBDA, HTTP, NONE, RELATIONAL_DATABASE, AMAZON_EVENTBRIDGE
/dataSources: contains invalid data source definitions"
`;

exports[`DynamoDB Invalid should validate: Empty config 1`] = `
"/dataSources/myDynamoSource1/config: must have required property 'tableName'
/dataSources: contains invalid data source definitions"
`;

exports[`DynamoDB Invalid should validate: Invalid config 1`] = `
"/dataSources/myDynamoSource1/config/tableName: must be a string or a CloudFormation intrinsic function
/dataSources/myDynamoSource1/config/useCallerCredentials: must be boolean
/dataSources/myDynamoSource1/config/serviceRoleArn: must be a string or a CloudFormation intrinsic function
/dataSources/myDynamoSource1/config/region: must be a string or a CloudFormation intrinsic function
/dataSources/myDynamoSource1/config/iamRoleStatements/0: must be a valid IAM role statement
/dataSources/myDynamoSource1/config/versioned: must be boolean
/dataSources/myDynamoSource1/config/deltaSyncConfig: must have required property 'deltaSyncTableName'
/dataSources/myDynamoSource1/config/deltaSyncConfig/baseTableTTL: must be integer
/dataSources/myDynamoSource1/config/deltaSyncConfig/deltaSyncTableTTL: must be integer
/dataSources: contains invalid data source definitions"
`;

exports[`DynamoDB Invalid should validate: Missing config 1`] = `
"/dataSources/myDynamoSource1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;

exports[`EventBridge Invalid should not validate: Empty config 1`] = `
"/dataSources/myEventBridgeSource1/config: must have required property 'eventBusArn'
/dataSources: contains invalid data source definitions"
`;

exports[`EventBridge Invalid should not validate: Invalid config 1`] = `
"/dataSources/myEventBridgeSource1/config/eventBusArn: must be a string or a CloudFormation intrinsic function
/dataSources: contains invalid data source definitions"
`;

exports[`EventBridge Invalid should not validate: Missing config 1`] = `
"/dataSources/myEventBridgeSource1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;

exports[`HTTP Invalid should validate: Empty config 1`] = `
"/dataSources/http1/config: must have required property 'endpoint'
/dataSources: contains invalid data source definitions"
`;

exports[`HTTP Invalid should validate: Invalid config 1`] = `
"/dataSources/http1/config/endpoint: must be a string or a CloudFormation intrinsic function
/dataSources/http1/config/authorizationConfig/authorizationType: must be AWS_IAM
/dataSources/http1/config/authorizationConfig/awsIamConfig/signingRegion: must be a string or a CloudFormation intrinsic function
/dataSources/http1/config/authorizationConfig/awsIamConfig/signingServiceName: must be a string or a CloudFormation intrinsic function
/dataSources: contains invalid data source definitions"
`;

exports[`HTTP Invalid should validate: Missing config 1`] = `
"/dataSources/http1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Empty config 1`] = `
"/dataSources/myLambda1/config: must specify functionName, functionArn or function (all exclusives)
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Invalid config 1`] = `
"/dataSources/myLambda1/config: must specify functionName, functionArn or function (all exclusives)
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Invalid embedded function 1`] = `
"/dataSources/myLambda1/config: must specify functionName, functionArn or function (all exclusives)
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Invalid functionArn 1`] = `
"/dataSources/myLambda1/config/functionArn: must be a string or a CloudFormation intrinsic function
/dataSources/myLambda1/config: must specify functionName, functionArn or function (all exclusives)
/dataSources/myLambda1/config/functionArn: must be a string or a CloudFormation intrinsic function
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Invalid functionName 1`] = `
"/dataSources/myLambda1/config: must specify functionName, functionArn or function (all exclusives)
/dataSources/myLambda1/config/functionName: must be string
/dataSources: contains invalid data source definitions"
`;

exports[`Lambda Invalid should validate: Missing config 1`] = `
"/dataSources/myLambda1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;

exports[`OpenSearch Invalid should validate: Empty config 1`] = `
"/dataSources/openSearch1/config: must have a endpoint or domain (but not both)
/dataSources: contains invalid data source definitions"
`;

exports[`OpenSearch Invalid should validate: Invalid config 1`] = `
"/dataSources/openSearch1/config/endpoint: must be a string or a CloudFormation intrinsic function
/dataSources/openSearch1/config: must have a endpoint or domain (but not both)
/dataSources/openSearch1/config/endpoint: must be a string or a CloudFormation intrinsic function
/dataSources/openSearch1/config/region: must be a string or a CloudFormation intrinsic function
/dataSources: contains invalid data source definitions"
`;

exports[`OpenSearch Invalid should validate: Missing config 1`] = `
"/dataSources/openSearch1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;

exports[`RelationalDb Invalid should validate: Empty config 1`] = `
"/dataSources/http1/config: must have required property 'endpoint'
/dataSources: contains invalid data source definitions"
`;

exports[`RelationalDb Invalid should validate: Invalid config 1`] = `
"/dataSources/http1/config/endpoint: must be a string or a CloudFormation intrinsic function
/dataSources/http1/config/authorizationConfig/authorizationType: must be AWS_IAM
/dataSources/http1/config/authorizationConfig/awsIamConfig/signingRegion: must be a string or a CloudFormation intrinsic function
/dataSources/http1/config/authorizationConfig/awsIamConfig/signingServiceName: must be a string or a CloudFormation intrinsic function
/dataSources: contains invalid data source definitions"
`;

exports[`RelationalDb Invalid should validate: Missing config 1`] = `
"/dataSources/http1: must have required property 'config'
/dataSources: contains invalid data source definitions"
`;
